build.sbt

```diff
@@ -43,12 +49,17 @@ libraryDependencies ++= Seq(
   "io.circe" %% "circe-parser" % circeVersion,
   "org.apache.spark" %% "spark-sql" % sparkVersion % Provided,
   "org.apache.spark" %% "spark-hive" % sparkVersion % Provided,
+  "com.hortonworks.hive" %% "hive-warehouse-connector" % "1.0.0.3.1.0.53-1" % "compileonly",
```

Comment:
According to the [SO discussion](https://stackoverflow.com/questions/21515325/add-a-compile-time-only-dependency-in-sbt), the `compileonly` scope suggests that there is no runtime dependency on this library, which is not true in this case (if `useHWC` is set to `true`, there will be a runtime dependency on `hive-warehouse-connector`). Do you think we should use the `provided` scope instead?

Reply:
Yes, I kind of abused this.  Essentially, what I want at compile time is something like a stub class that defines a `HiveWarehouseSession` and then at runtime supply the dependency that provides the implementation.  The reason for this is, provided the interface for `HiveWarehouseSession` is not changing among different versions, the user can supply whatever hwc jar is appropriate for their system.  The different builds of hwc depend on the different versions of hive.  Maybe as @dougb mentioned, I can an investigate an approach [reflection](https://docs.scala-lang.org/overviews/reflection/overview.html).

Reply:
I believe what you are describing better fits the structural programming concept in Scala. I don't think that would really be necessary.

You should be able to compile directly against HDP's `HiveWarehouseSession` unless you expect users to want to use some other variation of HWC (with a different class name or package name?)

"Provided" and "CompileOnly" should be basically the same except "Provided" leaves a transitive dependency indicator marked as "provided" in the published `pom.xml`.  I don't know of other differences. I don't think `spark-submit` will retrieve transitive dependencies marked "provided".

Reply:
Scala's structural programming allows you to define an after-the-fact interface and program against it, like this:
```scala
scala> type SomethingWithName = AnyRef {
     |   def name: String
     | }
defined type alias SomethingWithName

scala> def printName(named: SomethingWithName) = println(named.name)
printName: (named: SomethingWithName)Unit

scala> case class Named(name: String)
defined class Named

scala> printName(Named("Chris"))
Chris

scala> class SomeClass {
     |   def name = "Chris"
     | }
defined class SomeClass

scala> printName(new SomeClass)
Chris
```
This way the original compiler does not have to know about your defined interface.

The current implementation does use reflection. I don't know how scala 3.0 / dotty do this, or if they kept the feature.

Reply:
I don't know if I agree with the reasoning on `provided`.  That would indicate to me that the runtime environment should provide the **exact version** of the dependency.  I found [this blog post](https://blog.gradle.org/introducing-compile-only-dependencies) and these two use cases sum up what I had in mind:

> - Dependencies required at compile time but required at runtime only when using certain features, a.k.a. optional dependencies;
> - Dependencies whose API is required at compile time but whose implementation is to be provided by a consuming library, application or runtime environment. 

So the first bullet, I think, is apparent.  And for the second, I basically intended to say that the runtime environment can provide anything that implements the same API as `com.hortonworks.hive:hive-warehouse-connector:1.0.0.3.1.0.53-1`.

Does this explanation help?  

The only other thing I will add, and @samratmitra-0812 encountered this as well, is when the `hive-warehouse-connector` is marked as provided a whole host of dependency conflicts are introduced.